src/host-config/index.ts

```diff
@@ -57,75 +62,149 @@ export function parse(
+	const dns = {
+		...currentConf.dns,
+		...targetConf.dns,
+	};
+	patchedConf.dns = dns as DnsConfig;
+
+	// Delete dns config if null
+	if (targetConf.dns === null) {
+		delete patchedConf.dns;
 	}
+
+	// This method assumes that currentConf is a full ProxyConfig
+	// for backwards compatibility, so patchedConf.redsocks should
+	// also be a full ProxyConfig.
+	const redsocks = {
+		...currentConf.redsocks,
+		...targetConf.redsocks,
+	};
+	patchedConf.redsocks = redsocks as ProxyConfig;
+
+	// Delete redsocks and dns configs if redsocks is not configured,
+	// because dns config depends on redsocks to work
+	if (
+		targetConf.redsocks === null ||
+		(typeof targetConf.redsocks === 'object' &&
+			Object.keys(targetConf.redsocks).length === 0)
+	) {
+		delete patchedConf.redsocks;
+		delete patchedConf.dns;
+	}
+
 	return patchedConf;
 }
 
 export async function patch(
-	conf: HostConfiguration | LegacyHostConfiguration,
+	targetConf: HostConfiguration | LegacyHostConfiguration,
 	force: boolean = false,
 ): Promise<void> {
 	const apps = await applicationManager.getCurrentApps();
 	const appIds = Object.keys(apps).map((strId) => parseInt(strId, 10));
 
-	if (conf.network.hostname != null) {
-		await setHostname(conf.network.hostname);
+	const { network: target } = targetConf;
+
+	if (target.hostname != null) {
+		await setHostname(target.hostname);
 	}
 
-	if (conf.network.proxy != null) {
-		const targetConf = conf.network.proxy;
+	if (target.proxy != null || target.dns != null) {
 		// It's possible for appIds to be an empty array, but patch shouldn't fail
 		// as it's not dependent on there being any running user applications.
 		return updateLock.lock(appIds, { force }, async () => {
-			const proxyConf = await readProxy();
-			let currentConf: ProxyConfig | undefined = undefined;
-			if (proxyConf) {
-				delete proxyConf.noProxy;
-				currentConf = proxyConf;
+			// Get current proxy config
+			const current = await readProxy();
+			const currentConf: Partial<RedsocksConfig> = {};
+			if (current) {
+				const { proxy, dns } = current;
+				if (proxy) {
+					delete proxy.noProxy;
+					currentConf.redsocks = proxy;
+				}
+				if (dns) {
+					currentConf.dns = dns;
+				}
+			}
+
+			const toPatch: {
+				redsocks?: Partial<ProxyConfig> | null;
+				dns?: DnsConfig | null;
+			} = {};
+
+			// Extract target proxy from input
+			if (target.proxy) {
+				// target.proxy could be Record<string, any> due to legacy support,
+				// so only extract only defined fields that are part of ProxyConfig
+				const { type, ip, port, login, password } = target.proxy;
+				const definedConf = Object.fromEntries(
+					Object.entries({ type, ip, port, login, password }).filter(
+						([, value]) => value,
+					),
+				) as ProxyConfig;
+				if (Object.keys(definedConf).length > 0) {
+					toPatch.redsocks = definedConf;
+				} else {
+					// Mark for deletion if target proxy is empty
+					toPatch.redsocks = null;
+				}
+			}
```

Comment:
Why is this chunk of code necessary now and it wasn't before?

Reply:
Simplified per code review, so this chunk of code is no longer here